### PR TITLE
fix double backward for half softmax/logsoftmax

### PR DESCRIPTION
**Pull request by ngimel**
Fix for #17261, @SsnL do you have tests for it in your other PR? If not, I'll add to this. Example from #17261 now does not error out (and same for log_softmax).

### CODE REVIEW

**Comment by SsnL** [comment]
Thanks! Maybe removing [this line](https://github.com/pytorch/pytorch/blob/c02e2ff0b0a67e5ee39081b82da1316369520f12/test/test_nn.py#L2146) is sufficient for a test.
https://github.com/pytorch/pytorch/blob/c02e2ff0b0a67e5ee39081b82da1316369520f12/test/test_nn.py#L2146

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1206,8 +1206,8 @@
   self: log_sigmoid_double_backward(grad * grad_output, self)
 
 - name: _log_softmax_backward_data(Tensor grad_output, Tensor output, int64_t dim, Tensor self)
-  grad_output: grad - (grad * output.exp()).sum(dim, true)
-  self: log_softmax_double_backward(grad, grad_output, dim, output).type_as(self)
+  grad_output: grad.type_as(output) - (grad.type_as(output) * output.exp()).sum(dim, true)
```
Comment (SsnL) [inline tools/autograd/derivatives.yaml]: Could you use `.to(output.dtype())` instead? I know it's not a problem here, but `type_as` ignores devices index and I think we should generally avoid using it.

Comment (ngimel) [reply tools/autograd/derivatives.yaml]: Also replace type_as in clamp derivative.

Comment (SsnL) [reply tools/autograd/derivatives.yaml]: Thanks!

**Review by ssnl — APPROVED** [review]
This LGTM, but let's wait until CI comes up again.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (SsnL) [reply tools/autograd/derivatives.yaml]: I also wonder if we can avoid calculating `grad.type_as(output)` twice, but I guess that requires putting this into a function, and can be left as a future optimization.

Comment (ngimel) [reply tools/autograd/derivatives.yaml]: Yeah, I've noticed it, but given that it's really a corner case (in most cases type_as will be a no-op), decided to leave it as is for now

Comment (SsnL) [reply tools/autograd/derivatives.yaml]: SGTM

**Comment by ngimel** [comment]
I don't think windows build failure is related, it's coming from thrust iterator
```
15:10:39          C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.2/include\thrust/iterator/iterator_facade.h(517): error C2065: '__T0': undeclared identifier [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\build\caffe2\caffe2_gpu.vcxproj]
```
